In [31]:
import tensorflow as tf

In [32]:
import numpy as np

In [33]:
mnist=tf.keras.datasets.mnist

In [34]:
from sklearn.model_selection import KFold

In [35]:
kf=KFold(n_splits=5)

In [36]:
kf.get_n_splits()

5

In [37]:
(x_train, y_train),(x_test,y_test)=mnist.load_data()

In [38]:
x_train,x_test=x_train/255.0,x_test/255.0

In [39]:
model=tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [40]:
predictions=model(x_train[:5]).numpy()

In [41]:
predictions

array([[-0.18716869, -0.2359003 ,  0.478733  , -0.00483475, -0.04381356,
        -0.12678167,  0.3640622 ,  0.20057632,  0.2866352 , -0.19277745],
       [ 0.69342905, -0.14399509, -0.16396268, -0.49224535,  0.04645634,
        -0.16981356,  0.47967428, -0.18369412,  0.05232172, -0.17309113],
       [-0.24755739,  0.20746441,  0.16693929,  0.19467525,  0.30955532,
         0.42733932,  0.6105425 , -0.22113898,  0.47946236, -0.28529316],
       [ 0.18391863,  0.23735173,  0.10937361,  0.11730405, -0.07101984,
         0.3979329 ,  0.2738886 , -0.5452429 ,  0.53214455, -0.11001669],
       [ 0.05344913,  0.2783614 ,  0.42362773, -0.69517916,  0.09537707,
         0.40336055, -0.38962963, -0.5481131 ,  0.4695988 , -0.52521706]],
      dtype=float32)

In [42]:
tf.nn.softmax(predictions).numpy()

array([[0.07619265, 0.07256868, 0.14828955, 0.09143232, 0.08793695,
        0.08093546, 0.1322238 , 0.11228161, 0.12237241, 0.0757665 ],
       [0.18944068, 0.08199447, 0.08037347, 0.05788172, 0.09919655,
        0.07990459, 0.15298213, 0.07880314, 0.09978008, 0.07964313],
       [0.06340041, 0.09993222, 0.09596342, 0.09866231, 0.11067335,
        0.12450764, 0.14954095, 0.06509767, 0.13116947, 0.06105253],
       [0.10340679, 0.10908243, 0.09597864, 0.09674282, 0.08013656,
        0.12808383, 0.11314166, 0.04987447, 0.14648113, 0.07707164],
       [0.10099223, 0.12646376, 0.14623606, 0.04777083, 0.10531664,
        0.1433021 , 0.06484273, 0.05533919, 0.1531156 , 0.05662085]],
      dtype=float32)

In [43]:
loss_fn=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [44]:
loss_fn(y_train[:5],predictions).numpy()

2.293197

In [45]:
model.compile(optimizer='adam',loss=loss_fn,metrics=['accuracy'])

In [46]:

for train_index, test_index in kf.split(x_train):
    x_trains, x_tests=x_train[train_index],x_train[test_index]
    y_trains, y_tests=y_train[train_index], y_train[test_index]
    
    model.fit(x_trains,y_trains, validation_data=(x_tests,y_tests), epochs=5)
    
    
    

Epoch 1/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.3264 - accuracy: 0.9041 - val_loss: 0.1599 - val_accuracy: 0.9565
Epoch 2/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.1580 - accuracy: 0.9532 - val_loss: 0.1163 - val_accuracy: 0.9653
Epoch 3/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.1175 - accuracy: 0.9643 - val_loss: 0.0964 - val_accuracy: 0.9715
Epoch 4/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.0956 - accuracy: 0.9706 - val_loss: 0.0874 - val_accuracy: 0.9728
Epoch 5/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.0824 - accuracy: 0.9739 - val_loss: 0.0795 - val_accuracy: 0.9759
Epoch 1/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.0791 - accuracy: 0.9769 - val_loss: 0.0494 - val_accuracy: 0.9847
Epoch 2/5
1500/1500 [==============================] - 2s 1ms/step - loss: 0.0685 - accuracy: 0.9787 - val_loss: 0.0558 - val_accuracy: 0.9813

In [47]:
model.evaluate(x_test,y_test,verbose=2)

313/313 - 0s - loss: 0.0807 - accuracy: 0.9809


[0.08066778630018234, 0.98089998960495]

The end accuracy was 0.9809 compared to 0.9496 when not using KFold.

The predictions for the first 5: